File: packages/neptune_flutter_ui/lib/src/widgets/neptune_form_fields.dart
// © 2026 Neptune.Fintech (neptune.ly) · Neptune Odyssey Community License v1.0
//
// Branded form inputs (web `<npt-text-field>`, `<npt-select>`,
// `<npt-stepper>`, `<npt-date-field>`): themed text/select/stepper/date
// controls that match the brand decoration of the money inputs — filled
// surfaceContainerHighest, shape.rSm corners, primary focus ring, error
// states. Theme-only (no literal colours/radii/fonts), RTL-safe.

import 'package:flutter/material.dart';

import '../theme/extensions.dart';

/// A labelled, branded text input (web `<npt-text-field>`).
///
/// Renders an optional [label] (labelMedium, onSurfaceVariant) above a filled
/// [TextField] (surfaceContainerHighest, [NptShape.rSm] corners: no border at
/// rest, a 2px primary ring on focus, the error colour when [errorText] is
/// set), with optional prefix/suffix icons and helper/error text below.
/// Theme-only, RTL-safe.
class NeptuneTextField extends StatelessWidget {
  /// Optional field label shown above the input.
  final String? label;

  /// Placeholder text shown when the field is empty.
  final String? hint;

  /// Supporting text shown below the field when there is no [errorText].
  final String? helperText;

  /// Error text shown below the field; also recolours the border + label.
  final String? errorText;

  /// Optional leading icon inside the field.
  final IconData? prefixIcon;

  /// Optional trailing icon inside the field.
  final IconData? suffixIcon;

  /// Whether to obscure the value (e.g. passwords).
  final bool obscureText;

  /// Whether the field accepts input.
  final bool enabled;

  /// The editing controller for the field.
  final TextEditingController? controller;

  /// Called whenever the value changes.
  final ValueChanged<String>? onChanged;

  /// Called when the user submits (e.g. taps done).
  final ValueChanged<String>? onSubmitted;

  /// The keyboard type to request.
  final TextInputType? keyboardType;

  /// Optional maximum length (shows the counter when set).
  final int? maxLength;

  const NeptuneTextField({
    super.key,
    this.label,
    this.hint,
    this.helperText,
    this.errorText,
    this.prefixIcon,
    this.suffixIcon,
    this.obscureText = false,
    this.enabled = true,
    this.controller,
    this.onChanged,
    this.onSubmitted,
    this.keyboardType,
    this.maxLength,
  });

  @override
  Widget build(BuildContext context) {
    final scheme = Theme.of(context).colorScheme;
    final shape = Theme.of(context).extension<NptShape>()!;
    final text = Theme.of(context).textTheme;

    final hasError = errorText != null;
    final labelColor = hasError
        ? scheme.error
        : (enabled ? scheme.onSurfaceVariant : scheme.onSurfaceVariant.withValues(alpha: 0.5));

    OutlineInputBorder border(Color color, double width) => OutlineInputBorder(
          borderRadius: shape.rSm,
          borderSide: width == 0 && color == scheme.surface
              ? BorderSide.none
              : BorderSide(color: color, width: width),
        );

    return Column(
      crossAxisAlignment: CrossAxisAlignment.start,
      mainAxisSize: MainAxisSize.min,
      children: [
        if (label != null) ...[
          Text(
            label!,
            style: text.labelMedium?.copyWith(color: labelColor),
          ),
          const SizedBox(height: 6),
        ],
        TextField(
          controller: controller,
          enabled: enabled,
          obscureText: obscureText,
          onChanged: onChanged,
          onSubmitted: onSubmitted,
          keyboardType: keyboardType,
          maxLength: maxLength,
          cursorColor: scheme.primary,
          style: text.bodyLarge?.copyWith(color: scheme.onSurface),
          decoration: InputDecoration(
            isDense: true,
            filled: true,
            fillColor: enabled
                ? scheme.surfaceContainerHighest
                : scheme.surfaceContainerHighest.withValues(alpha: 0.5),
            hintText: hint,
            hintStyle: text.bodyLarge?.copyWith(
              color: scheme.onSurfaceVariant.withValues(alpha: 0.6),
            ),
            prefixIcon: prefixIcon == null
                ? null
                : Icon(prefixIcon, size: 20, color: scheme.onSurfaceVariant),
            suffixIcon: suffixIcon == null
                ? null
                : Icon(suffixIcon, size: 20, color: scheme.onSurfaceVariant),
            counterText: maxLength == null ? '' : null,
            contentPadding: const EdgeInsetsDirectional.symmetric(
              horizontal: 16,
              vertical: 14,
            ),
            // No visible border at rest; primary ring on focus; error tint.
            border: OutlineInputBorder(
              borderRadius: shape.rSm,
              borderSide: BorderSide.none,
            ),
            enabledBorder: hasError
                ? border(scheme.error, 1)
                : OutlineInputBorder(
                    borderRadius: shape.rSm,
                    borderSide: BorderSide.none,
                  ),
            disabledBorder: OutlineInputBorder(
              borderRadius: shape.rSm,
              borderSide: BorderSide(color: scheme.outlineVariant),
            ),
            focusedBorder: border(hasError ? scheme.error : scheme.primary, 2),
            errorBorder: border(scheme.error, 1),
            focusedErrorBorder: border(scheme.error, 2),
            // We render our own helper/error row below for full style control.
            errorStyle: const TextStyle(height: 0, fontSize: 0),
          ),
        ),
        if (hasError || helperText != null) ...[
          const SizedBox(height: 6),
          Text(
            errorText ?? helperText!,
            style: text.bodySmall?.copyWith(
              color: hasError ? scheme.error : scheme.onSurfaceVariant,
            ),
          ),
        ],
      ],
    );
  }
}

/// One option in a [NeptuneSelect]: a [value], its [label], and an optional
/// leading [icon].
class NeptuneSelectOption<T> {
  /// The value carried by this option.
  final T value;

  /// The human-readable label.
  final String label;

  /// Optional leading icon shown in the menu + the closed field.
  final IconData? icon;

  const NeptuneSelectOption({
    required this.value,
    required this.label,
    this.icon,
  });
}

/// A labelled, branded dropdown select (web `<npt-select>`).
///
/// A themed [DropdownButtonFormField] whose decoration matches
/// [NeptuneTextField] (filled surfaceContainerHighest, [NptShape.rSm] corners,
/// primary focus ring). Menu items show an optional [NeptuneSelectOption.icon]
/// before the label. Theme-only, RTL-safe.
class NeptuneSelect<T> extends StatelessWidget {
  /// Optional field label shown above the select.
  final String? label;

  /// Placeholder shown when no value is selected.
  final String? hint;

  /// The selectable options.
  final List<NeptuneSelectOption<T>> options;

  /// The currently selected value.
  final T? value;

  /// Called when the selection changes.
  final ValueChanged<T?>? onChanged;

  /// Whether the select accepts input.
  final bool enabled;

  const NeptuneSelect({
    super.key,
    this.label,
    this.hint,
    required this.options,
    this.value,
    this.onChanged,
    this.enabled = true,
  });

  @override
  Widget build(BuildContext context) {
    final scheme = Theme.of(context).colorScheme;
    final shape = Theme.of(context).extension<NptShape>()!;
    final text = Theme.of(context).textTheme;

    OutlineInputBorder border(Color color, double width) => OutlineInputBorder(
          borderRadius: shape.rSm,
          borderSide: BorderSide(color: color, width: width),
        );

    return Column(
      crossAxisAlignment: CrossAxisAlignment.start,
      mainAxisSize: MainAxisSize.min,
      children: [
        if (label != null) ...[
          Text(
            label!,
            style: text.labelMedium?.copyWith(
              color: enabled
                  ? scheme.onSurfaceVariant
                  : scheme.onSurfaceVariant.withValues(alpha: 0.5),
            ),
          ),
          const SizedBox(height: 6),
        ],
        DropdownButtonFormField<T>(
          initialValue: value,
          onChanged: enabled ? onChanged : null,
          isExpanded: true,
          borderRadius: shape.rSm,
          icon: Icon(Icons.arrow_drop_down, color: scheme.onSurfaceVariant),
          dropdownColor: scheme.surfaceContainerHigh,
          style: text.bodyLarge?.copyWith(color: scheme.onSurface),
          hint: hint == null
              ? null
              : Text(
                  hint!,
                  style: text.bodyLarge?.copyWith(
                    color: scheme.onSurfaceVariant.withValues(alpha: 0.6),
                  ),
                ),
          decoration: InputDecoration(
            isDense: true,
            filled: true,
            fillColor: enabled
                ? scheme.surfaceContainerHighest
                : scheme.surfaceContainerHighest.withValues(alpha: 0.5),
            contentPadding: const EdgeInsetsDirectional.symmetric(
              horizontal: 16,
              vertical: 14,
            ),
            border: OutlineInputBorder(
              borderRadius: shape.rSm,
              borderSide: BorderSide.none,
            ),
            enabledBorder: OutlineInputBorder(
              borderRadius: shape.rSm,
              borderSide: BorderSide.none,
            ),
            disabledBorder: OutlineInputBorder(
              borderRadius: shape.rSm,
              borderSide: BorderSide(color: scheme.outlineVariant),
            ),
            focusedBorder: border(scheme.primary, 2),
          ),
          items: [
            for (final option in options)
              DropdownMenuItem<T>(
                value: option.value,
                child: Row(
                  mainAxisSize: MainAxisSize.min,
                  children: [
                    if (option.icon != null) ...[
                      Icon(option.icon, size: 20, color: scheme.onSurfaceVariant),
                      const SizedBox(width: 12),
                    ],
                    Flexible(
                      child: Text(
                        option.label,
                        overflow: TextOverflow.ellipsis,
                        style: text.bodyLarge?.copyWith(color: scheme.onSurface),
                      ),
                    ),
                  ],
                ),
              ),
          ],
        ),
      ],
    );
  }
}

/// A branded numeric +/- stepper (web `<npt-stepper>`).
///
/// A pill row: a minus [IconButton], the centred [value] (tabular, titleMedium)
/// and a plus [IconButton]. Minus disables at [min], plus disables at [max].
/// Controls are ≥48dp. Theme-only, RTL-safe.
class NeptuneStepperInput extends StatelessWidget {
  /// The current value.
  final int value;

  /// The smallest allowed value (inclusive).
  final int min;

  /// The largest allowed value (inclusive).
  final int max;

  /// The increment/decrement applied per tap.
  final int step;

  /// Called with the clamped new value when +/- is tapped.
  final ValueChanged<int>? onChanged;

  /// Optional label shown above the stepper.
  final String? label;

  const NeptuneStepperInput({
    super.key,
    required this.value,
    this.min = 0,
    this.max = 9999,
    this.step = 1,
    this.onChanged,
    this.label,
  });

  @override
  Widget build(BuildContext context) {
    final scheme = Theme.of(context).colorScheme;
    final shape = Theme.of(context).extension<NptShape>()!;
    final text = Theme.of(context).textTheme;

    final canDecrement = onChanged != null && value > min;
    final canIncrement = onChanged != null && value < max;
    final radius = BorderRadius.circular(shape.full);

    void emit(int next) => onChanged?.call(next.clamp(min, max));

    Widget control({
      required IconData icon,
      required bool enabled,
      required VoidCallback onTap,
      required String tooltip,
    }) {
      final fg = enabled ? scheme.onSurface : scheme.onSurfaceVariant.withValues(alpha: 0.4);
      return IconButton(
        onPressed: enabled ? onTap : null,
        tooltip: tooltip,
        iconSize: 20,
        constraints: const BoxConstraints(minWidth: 48, minHeight: 48),
        icon: Icon(icon, color: fg),
      );
    }

    return Column(
      crossAxisAlignment: CrossAxisAlignment.start,
      mainAxisSize: MainAxisSize.min,
      children: [
        if (label != null) ...[
          Text(
            label!,
            style: text.labelMedium?.copyWith(color: scheme.onSurfaceVariant),
          ),
          const SizedBox(height: 6),
        ],
        Container(
          decoration: BoxDecoration(
            color: scheme.surfaceContainerHighest,
            borderRadius: radius,
            border: Border.all(color: scheme.outlineVariant),
          ),
          child: Row(
            mainAxisSize: MainAxisSize.min,
            children: [
              control(
                icon: Icons.remove,
                enabled: canDecrement,
                onTap: () => emit(value - step),
                tooltip: 'Decrease',
              ),
              ConstrainedBox(
                constraints: const BoxConstraints(minWidth: 40),
                child: Text(
                  '$value',
                  textAlign: TextAlign.center,
                  style: text.titleMedium?.copyWith(
                    color: scheme.onSurface,
                    fontFeatures: const [FontFeature.tabularFigures()],
                  ),
                ),
              ),
              control(
                icon: Icons.add,
                enabled: canIncrement,
                onTap: () => emit(value + step),
                tooltip: 'Increase',
              ),
            ],
          ),
        ),
      ],
    );
  }
}

/// A branded date field (web `<npt-date-field>`).
///
/// A tappable field with a calendar_today suffix that opens [showDatePicker]
/// (inheriting the active theme) and shows the selected [value] formatted as
/// `yyyy-MM-dd`. Decoration matches [NeptuneTextField]. Theme-only, RTL-safe.
class NeptuneDateField extends StatelessWidget {
  /// Optional field label shown above the input.
  final String? label;

  /// Placeholder shown when no date is selected.
  final String? hint;

  /// The currently selected date.
  final DateTime? value;

  /// Called with the picked date when the user confirms.
  final ValueChanged<DateTime>? onChanged;

  /// The earliest selectable date (defaults to 100 years ago).
  final DateTime? firstDate;

  /// The latest selectable date (defaults to 100 years ahead).
  final DateTime? lastDate;

  const NeptuneDateField({
    super.key,
    this.label,
    this.hint,
    this.value,
    this.onChanged,
    this.firstDate,
    this.lastDate,
  });

  /// Format a date as `yyyy-MM-dd` without depending on `intl`.
  static String _format(DateTime d) {
    final y = d.year.toString().padLeft(4, '0');
    final m = d.month.toString().padLeft(2, '0');
    final day = d.day.toString().padLeft(2, '0');
    return '$y-$m-$day';
  }

  Future<void> _pick(BuildContext context) async {
    final now = DateTime.now();
    final first = firstDate ?? DateTime(now.year - 100);
    final last = lastDate ?? DateTime(now.year + 100);
    var initial = value ?? now;
    if (initial.isBefore(first)) initial = first;
    if (initial.isAfter(last)) initial = last;
    final picked = await showDatePicker(
      context: context,
      initialDate: initial,
      firstDate: first,
      lastDate: last,
    );
    if (picked != null) onChanged?.call(picked);
  }

  @override
  Widget build(BuildContext context) {
    final scheme = Theme.of(context).colorScheme;
    final shape = Theme.of(context).extension<NptShape>()!;
    final text = Theme.of(context).textTheme;

    final hasValue = value != null;
    final display = hasValue ? _format(value!) : (hint ?? 'yyyy-MM-dd');

    return Column(
      crossAxisAlignment: CrossAxisAlignment.start,
      mainAxisSize: MainAxisSize.min,
      children: [
        if (label != null) ...[
          Text(
            label!,
            style: text.labelMedium?.copyWith(color: scheme.onSurfaceVariant),
          ),
          const SizedBox(height: 6),
        ],
        Material(
          color: scheme.surfaceContainerHighest,
          borderRadius: shape.rSm,
          clipBehavior: Clip.antiAlias,
          child: InkWell(
            onTap: () => _pick(context),
            customBorder: RoundedRectangleBorder(borderRadius: shape.rSm),
            child: ConstrainedBox(
              constraints: const BoxConstraints(minHeight: 48),
              child: Padding(
                padding: const EdgeInsetsDirectional.symmetric(
                  horizontal: 16,
                  vertical: 12,
                ),
                child: Row(
                  children: [
                    Expanded(
                      child: Text(
                        display,
                        style: text.bodyLarge?.copyWith(
                          color: hasValue
                              ? scheme.onSurface
                              : scheme.onSurfaceVariant.withValues(alpha: 0.6),
                          fontFeatures:
                              hasValue ? const [FontFeature.tabularFigures()] : null,
                        ),
                      ),
                    ),
                    const SizedBox(width: 12),
                    Icon(
                      Icons.calendar_today,
                      size: 20,
                      color: scheme.onSurfaceVariant,
                    ),
                  ],
                ),
              ),
            ),
          ),
        ),
      ],
    );
  }
}

File: packages/neptune_kmp_ui/odyssey-compose-ui/src/commonMain/kotlin/ly/neptune/odyssey/ui/components/NeptuneFormFields.kt
// © 2026 Neptune.Fintech (neptune.ly) · Neptune Odyssey Community License v1.0
//
// Branded form inputs. Web counterparts: `<npt-select>`, `<npt-stepper>`,
// `<npt-date-field>` · Flutter: neptune_form_fields.dart. All three wear the
// NeptuneTextField decoration — filled surface-container-highest on the brand
// `sm` corner, 2dp primary focus ring, outline-variant hairline when
// disabled. Theme-only, ≥48dp targets, RTL-safe.

package ly.neptune.odyssey.ui.components

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.defaultMinSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.text.BasicTextField
import androidx.compose.foundation.text.KeyboardOptions
import androidx.compose.material3.DropdownMenu
import androidx.compose.material3.DropdownMenuItem
import androidx.compose.material3.Icon
import androidx.compose.material3.LocalContentColor
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.CompositionLocalProvider
import androidx.compose.runtime.Immutable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableIntStateOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.focus.onFocusChanged
import androidx.compose.ui.graphics.SolidColor
import androidx.compose.ui.graphics.vector.ImageVector
import androidx.compose.ui.hapticfeedback.HapticFeedback
import androidx.compose.ui.layout.onSizeChanged
import androidx.compose.ui.platform.LocalDensity
import androidx.compose.ui.platform.LocalHapticFeedback
import androidx.compose.ui.semantics.Role
import androidx.compose.ui.text.AnnotatedString
import androidx.compose.ui.text.input.KeyboardType
import androidx.compose.ui.text.input.OffsetMapping
import androidx.compose.ui.text.input.TransformedText
import androidx.compose.ui.text.input.VisualTransformation
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import ly.neptune.odyssey.ui.glyphs.NptInputGlyphs
import ly.neptune.odyssey.ui.theme.NeptuneTheme
import ly.neptune.odyssey.ui.theme.NptFeedbackCue

/** One option in a [NeptuneSelect]: a [value], its [label], and an optional
 * leading [icon] slot (20dp, onSurfaceVariant tint) shown in the menu and in
 * the closed field. Flutter: `NeptuneSelectOption`. */
@Immutable
public class NeptuneSelectOption<T>(
    public val value: T,
    public val label: String,
    public val icon: (@Composable () -> Unit)? = null,
)

/**
 * A labelled, branded dropdown select.
 *
 * Web counterpart: `<npt-select>` · Flutter: `NeptuneSelect`.
 *
 * The closed field matches [NeptuneTextField] (filled
 * `surfaceContainerHighest`, brand `sm` corners, 2dp primary ring while the
 * menu is open, outline-variant hairline when disabled) with a caret suffix.
 * The menu is a field-width `surfaceContainerHigh` popup on the same `sm`
 * corner; items show the option's icon slot before its label (bodyLarge,
 * onSurface). [onChanged] null (or `enabled = false`) disables the control.
 */
@Composable
public fun <T> NeptuneSelect(
    options: List<NeptuneSelectOption<T>>,
    value: T?,
    onChanged: ((T) -> Unit)?,
    modifier: Modifier = Modifier,
    label: String? = null,
    hint: String? = null,
    enabled: Boolean = true,
) {
    val scheme = MaterialTheme.colorScheme
    val typography = MaterialTheme.typography
    val shape = NeptuneTheme.shape.rSm
    val feedback = NeptuneTheme.feedback
    val haptics: HapticFeedback = LocalHapticFeedback.current
    val density = LocalDensity.current
    val isOn = enabled && onChanged != null

    var expanded by remember { mutableStateOf(false) }
    var fieldWidthPx by remember { mutableIntStateOf(0) }

    val fill = if (isOn) {
        scheme.surfaceContainerHighest
    } else {
        scheme.surfaceContainerHighest.copy(alpha = 0.5f)
    }
    // No border at rest; primary ring while open; hairline when disabled —
    // the NeptuneTextField border table.
    val borderModifier = when {
        expanded && isOn -> Modifier.border(2.dp, scheme.primary, shape)
        !isOn -> Modifier.border(1.dp, scheme.outlineVariant, shape)
        else -> Modifier
    }
    val selected = options.firstOrNull { it.value == value }

    Column(modifier) {
        if (label != null) {
            Text(
                text = label,
                style = typography.labelMedium.copy(
                    color = if (isOn) {
                        scheme.onSurfaceVariant
                    } else {
                        scheme.onSurfaceVariant.copy(alpha = 0.5f)
                    },
                ),
            )
            Spacer(Modifier.height(6.dp))
        }
        Box {
            Row(
                modifier = Modifier
                    .fillMaxWidth()
                    .onSizeChanged { fieldWidthPx = it.width }
                    .defaultMinSize(minHeight = 48.dp)
                    .clip(shape)
                    .background(fill)
                    .then(borderModifier)
                    .clickable(enabled = isOn, role = Role.DropdownList) {
                        feedback.trigger(NptFeedbackCue.Tap, haptics)
                        expanded = true
                    }
                    .padding(horizontal = 16.dp, vertical = 14.dp),
                verticalAlignment = Alignment.CenterVertically,
            ) {
                if (selected?.icon != null) {
                    CompositionLocalProvider(LocalContentColor provides scheme.onSurfaceVariant) {
                        Box(Modifier.size(20.dp), contentAlignment = Alignment.Center) {
                            selected.icon.invoke()
                        }
                    }
                    Spacer(Modifier.width(12.dp))
                }
                Text(
                    text = selected?.label ?: hint.orEmpty(),
                    maxLines = 1,
                    overflow = TextOverflow.Ellipsis,
                    style = typography.bodyLarge.copy(
                        color = if (selected != null) {
                            scheme.onSurface
                        } else {
                            scheme.onSurfaceVariant.copy(alpha = 0.6f)
                        },
                    ),
                    modifier = Modifier.weight(1f),
                )
                Spacer(Modifier.width(12.dp))
                Icon(
                    NptInputGlyphs.caretDown,
                    contentDescription = null,
                    tint = scheme.onSurfaceVariant,
                    modifier = Modifier.size(24.dp),
                )
            }
            DropdownMenu(
                expanded = expanded,
                onDismissRequest = { expanded = false },
                modifier = Modifier.width(with(density) { fieldWidthPx.toDp() }),
                shape = shape,
                containerColor = scheme.surfaceContainerHigh,
            ) {
                options.forEach { option ->
                    DropdownMenuItem(
                        text = {
                            Text(
                                text = option.label,
                                maxLines = 1,
                                overflow = TextOverflow.Ellipsis,
                                style = typography.bodyLarge.copy(color = scheme.onSurface),
                            )
                        },
                        leadingIcon = option.icon?.let { icon ->
                            {
                                CompositionLocalProvider(
                                    LocalContentColor provides scheme.onSurfaceVariant,
                                ) {
                                    Box(
                                        Modifier.size(20.dp),
                                        contentAlignment = Alignment.Center,
                                    ) { icon() }
                                }
                            }
                        },
                        onClick = {
                            feedback.trigger(NptFeedbackCue.Tap, haptics)
                            expanded = false
                            onChanged?.invoke(option.value)
                        },
                    )
                }
            }
        }
    }
}

/**
 * A branded numeric +/- stepper: a pill row (`surfaceContainerHighest` on
 * the brand `full` corner with an outline-variant hairline) of a minus
 * control, the centred value (titleMedium, tabular figures) and a plus
 * control. Minus disables at [min], plus at [max]; both are 48dp targets.
 *
 * Web counterpart: `<npt-stepper>` · Flutter: `NeptuneStepperInput`.
 *
 * [onChanged] receives the clamped new value per tap; null disables the
 * control. RTL-safe (minus leads the reading direction, like the sources).
 */
@Composable
public fun NeptuneStepperInput(
    value: Int,
    onChanged: ((Int) -> Unit)?,
    modifier: Modifier = Modifier,
    min: Int = 0,
    max: Int = 9999,
    step: Int = 1,
    label: String? = null,
) {
    val scheme = MaterialTheme.colorScheme
    val typography = MaterialTheme.typography
    val shape = NeptuneTheme.shape.rFull

    val canDecrement = onChanged != null && value > min
    val canIncrement = onChanged != null && value < max
    fun emit(next: Int) {
        onChanged?.invoke(next.coerceIn(min, max))
    }

    Column(modifier) {
        if (label != null) {
            Text(
                text = label,
                style = typography.labelMedium.copy(color = scheme.onSurfaceVariant),
            )
            Spacer(Modifier.height(6.dp))
        }
        Row(
            modifier = Modifier
                .clip(shape)
                .background(scheme.surfaceContainerHighest)
                .border(1.dp, scheme.outlineVariant, shape),
            verticalAlignment = Alignment.CenterVertically,
        ) {
            StepperControl(
                glyph = NptInputGlyphs.minus,
                contentDescription = "Decrease",
                enabled = canDecrement,
                onTap = { emit(value - step) },
            )
            Text(
                text = NeptuneTheme.formatDigits(value.toString()),
                textAlign = TextAlign.Center,
                style = typography.titleMedium.copy(
                    color = scheme.onSurface,
                    fontFeatureSettings = "tnum",
                ),
                modifier = Modifier.defaultMinSize(minWidth = 40.dp),
            )
            StepperControl(
                glyph = NptInputGlyphs.plus,
                contentDescription = "Increase",
                enabled = canIncrement,
                onTap = { emit(value + step) },
            )
        }
    }
}

/** A 48dp round +/- control used by [NeptuneStepperInput]. */
@Composable
private fun StepperControl(
    glyph: ImageVector,
    contentDescription: String,
    enabled: Boolean,
    onTap: () -> Unit,
) {
    val scheme = MaterialTheme.colorScheme
    val feedback = NeptuneTheme.feedback
    val haptics: HapticFeedback = LocalHapticFeedback.current
    val tint = if (enabled) scheme.onSurface else scheme.onSurfaceVariant.copy(alpha = 0.4f)

    Box(
        modifier = Modifier
            .size(48.dp)
            .clip(NeptuneTheme.shape.rFull)
            .clickable(enabled = enabled, role = Role.Button) {
                feedback.trigger(NptFeedbackCue.Tap, haptics)
                onTap()
            },
        contentAlignment = Alignment.Center,
    ) {
        Icon(
            glyph,
            contentDescription = contentDescription,
            tint = tint,
            modifier = Modifier.size(20.dp),
        )
    }
}

/** Inserts the `yyyy-MM-dd` dashes over a digits-only field value. */
private object DateDashTransformation : VisualTransformation {
    override fun filter(text: AnnotatedString): TransformedText {
        val out = buildString {
            text.text.forEachIndexed { i, c ->
                if (i == 4 || i == 6) append('-')
                append(c)
            }
        }
        val mapping = object : OffsetMapping {
            override fun originalToTransformed(offset: Int): Int =
                offset + (if (offset > 4) 1 else 0) + (if (offset > 6) 1 else 0)

            override fun transformedToOriginal(offset: Int): Int =
                offset - (if (offset > 4) 1 else 0) - (if (offset > 7) 1 else 0)
        }
        return TransformedText(AnnotatedString(out), mapping)
    }
}

/** Digits of [formatted] (max 8), dropping the dashes. */
private fun dateDigits(formatted: String): String =
    formatted.filter { it in '0'..'9' }.take(8)

/** Format [digits] (≤8) as a partial `yyyy-MM-dd` string. */
private fun formatDate(digits: String): String = buildString {
    digits.forEachIndexed { i, c ->
        if (i == 4 || i == 6) append('-')
        append(c)
    }
}

/**
 * A branded date field with formatted text entry: digits format as
 * `yyyy-MM-dd` while typing (no picker dependency — the Compose stand-in
 * for the Flutter widget's `showDatePicker` flow).
 *
 * Web counterpart: `<npt-date-field>` · Flutter: `NeptuneDateField`.
 *
 * Decoration matches [NeptuneTextField]: filled `surfaceContainerHighest`
 * on the brand `sm` corner (2dp primary ring on focus, outline-variant
 * hairline when disabled), bodyLarge value with tabular figures, the
 * `yyyy-MM-dd` [hint] at 60% onSurfaceVariant and a 20dp calendar suffix.
 *
 * [value] and [onValueChange] carry the formatted (possibly partial)
 * `yyyy-MM-dd` string; [onComplete] fires once all eight digits are typed.
 */
@Composable
public fun NeptuneDateField(
    value: String,
    onValueChange: (String) -> Unit,
    modifier: Modifier = Modifier,
    label: String? = null,
    hint: String? = null,
    enabled: Boolean = true,
    onComplete: ((String) -> Unit)? = null,
) {
    val scheme = MaterialTheme.colorScheme
    val typography = MaterialTheme.typography
    val shape = NeptuneTheme.shape.rSm
    var focused by remember { mutableStateOf(false) }

    val digits = dateDigits(value)
    val hasValue = digits.isNotEmpty()
    val fill = if (enabled) {
        scheme.surfaceContainerHighest
    } else {
        scheme.surfaceContainerHighest.copy(alpha = 0.5f)
    }
    val borderModifier = when {
        focused && enabled -> Modifier.border(2.dp, scheme.primary, shape)
        !enabled -> Modifier.border(1.dp, scheme.outlineVariant, shape)
        else -> Modifier
    }
    val textStyle = typography.bodyLarge.copy(
        color = scheme.onSurface,
        fontFeatureSettings = "tnum",
    )

    Column(modifier) {
        if (label != null) {
            Text(
                text = label,
                style = typography.labelMedium.copy(color = scheme.onSurfaceVariant),
            )
            Spacer(Modifier.height(6.dp))
        }
        BasicTextField(
            value = digits,
            onValueChange = { raw ->
                val clean = dateDigits(raw)
                if (clean != digits) {
                    val formatted = formatDate(clean)
                    onValueChange(formatted)
                    if (clean.length == 8) onComplete?.invoke(formatted)
                }
            },
            modifier = Modifier
                .fillMaxWidth()
                .onFocusChanged { focused = it.isFocused },
            enabled = enabled,
            textStyle = textStyle,
            cursorBrush = SolidColor(scheme.primary),
            visualTransformation = DateDashTransformation,
            keyboardOptions = KeyboardOptions(keyboardType = KeyboardType.Number),
            singleLine = true,
            decorationBox = { innerTextField ->
                Row(
                    modifier = Modifier
                        .defaultMinSize(minHeight = 48.dp)
                        .clip(shape)
                        .background(fill)
                        .then(borderModifier)
                        .padding(horizontal = 16.dp, vertical = 12.dp),
                    verticalAlignment = Alignment.CenterVertically,
                ) {
                    Box(Modifier.weight(1f)) {
                        if (!hasValue) {
                            Text(
                                text = hint ?: "yyyy-MM-dd",
                                maxLines = 1,
                                overflow = TextOverflow.Ellipsis,
                                style = typography.bodyLarge.copy(
                                    color = scheme.onSurfaceVariant.copy(alpha = 0.6f),
                                ),
                            )
                        }
                        innerTextField()
                    }
                    Spacer(Modifier.width(12.dp))
                    Icon(
                        NptInputGlyphs.calendar,
                        contentDescription = null,
                        tint = scheme.onSurfaceVariant,
                        modifier = Modifier.size(20.dp),
                    )
                }
            },
        )
    }
}
